core/src/abstractions.rs

```diff
@@ -30,9 +32,14 @@ impl MeteredSemaphore {
         res
     }
 
-    /// Adds just one permit
+    /// Adds just one permit. Will not add if already at the initial/max capacity.
     pub fn add_permit(&self) {
-        self.sem.add_permits(1);
-        (self.record_fn)(&self.metrics_ctx, self.sem.available_permits());
+        if self.sem.available_permits() < self.max_permits {
```

Comment:
Isn't this racy?

Reply:
Shouldn't be, since acquire is async and this is happening in sync code. This is effectively atomic from the viewpoint of someone trying to acquire.